pandas/core/reshape/encoding.py

```diff
@@ -316,3 +318,202 @@ def get_empty_frame(data) -> DataFrame:
             dummy_mat = dummy_mat[:, 1:]
             dummy_cols = dummy_cols[1:]
         return DataFrame(dummy_mat, index=index, columns=dummy_cols)
+
+
+def from_dummies(
+    data: DataFrame,
+    sep: None | str = None,
+    base_category: None | Hashable | dict[str, Hashable] = None,
+) -> DataFrame:
+    """
+    Create a categorical `DataFrame` from a `DataFrame` of dummy variables.
+
+    Inverts the operation performed by :func:`~pandas.get_dummies`.
+
+    Parameters
+    ----------
+    data : DataFrame
+        Data which contains dummy-coded variables.
+    sep : str, default None
+        Separator used in the column names of the dummy categories they are
+        character indicating the separation of the categorical names from the prefixes.
+        For example, if your column names are 'prefix_A' and 'prefix_B',
```

Comment:
Hmm this argument seems really specific to when DataFrame column labels are strings, but DataFrame column labels can be any hashable value (int, datetime, etc). I'd be weary of including this argument that is specific to a data type.

Reply:
I mirrored the way the different categories are created in `get_dummies` - in my understanding it makes sense in this case as most dummy DataFrame columns should be strings in order to match the categories as they require "modular" labels.

```bash
    >>> df
          col1_a  col1_b  col2_a  col2_b  col2_c
    0       1       0       0       1       0
    1       0       1       1       0       0
    2       1       0       0       0       1
```

Reply:
Ah! Okay that totally makes sense to have this argument then. Thanks.

Are there tests for when the column names are not strings?

Reply:
Yes, [for the 1d case where the column names are integers, floats, or mixed](https://github.com/pandas-dev/pandas/pull/41902/files#diff-7fa6fe95c70e67efabbb28f97c40969ad3d65ba54f183e586097b792190c1e94R253-R283)